src/platform/Linux/NetworkCommissioningWiFiDriver.cpp

```diff
@@ -123,40 +128,56 @@ void LinuxWiFiDriver::Shutdown()
-        ReturnErrorOnFailure(mStagingNetwork.clientIdentityKeypair->Serialize(serializedKeypair));
-        ReturnErrorOnFailure(
-            kvs.Put(kWifiClientIdentityKeypairKeyName, serializedKeypair.ConstBytes(), serializedKeypair.Length()));
-    }
-    else
+
+    for (int networkIndex = 0; networkIndex < currentNetId; networkIndex++)
     {
-        ReturnErrorOnFailure(IgnoreNotFound(kvs.Delete(kWifiNetworkIdentityKeyName)));
-        ReturnErrorOnFailure(IgnoreNotFound(kvs.Delete(kWifiClientIdentityKeyName)));
-        ReturnErrorOnFailure(IgnoreNotFound(kvs.Delete(kWifiClientIdentityKeypairKeyName)));
+        ReturnErrorOnFailure(kvs.Put(kWiFiSSIDKeyName, mStagingNetwork[networkIndex].ssid, mStagingNetwork[networkIndex].ssidLen));
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFI_PDC
+        if (mStagingNetwork[networkIndex].UsingPDC())
+        {
+            ReturnErrorOnFailure(IgnoreNotFound(kvs.Delete(kWiFiCredentialsKeyName)));
+            ReturnErrorOnFailure(kvs.Put(kWifiNetworkIdentityKeyName, mStagingNetwork[networkIndex].networkIdentity,
+                                         mStagingNetwork[networkIndex].networkIdentityLen));
+            ReturnErrorOnFailure(kvs.Put(kWifiClientIdentityKeyName, mStagingNetwork[networkIndex].clientIdentity,
+                                         mStagingNetwork[networkIndex].clientIdentityLen));
+
+            P256SerializedKeypair serializedKeypair;
+            ReturnErrorOnFailure(mStagingNetwork[networkIndex].clientIdentityKeypair->Serialize(serializedKeypair));
+            ReturnErrorOnFailure(
+                kvs.Put(kWifiClientIdentityKeypairKeyName, serializedKeypair.ConstBytes(), serializedKeypair.Length()));
+        }
+        else
+        {
+            ReturnErrorOnFailure(IgnoreNotFound(kvs.Delete(kWifiNetworkIdentityKeyName)));
+            ReturnErrorOnFailure(IgnoreNotFound(kvs.Delete(kWifiClientIdentityKeyName)));
+            ReturnErrorOnFailure(IgnoreNotFound(kvs.Delete(kWifiClientIdentityKeypairKeyName)));
 #else  // CHIP_DEVICE_CONFIG_ENABLE_WIFI_PDC
-    {
+        {
 #endif // CHIP_DEVICE_CONFIG_ENABLE_WIFI_PDC
-        ReturnErrorOnFailure(kvs.Put(kWiFiCredentialsKeyName, mStagingNetwork.credentials, mStagingNetwork.credentialsLen));
+            ReturnErrorOnFailure(kvs.Put(kWiFiCredentialsKeyName, mStagingNetwork[networkIndex].credentials,
+                                         mStagingNetwork[networkIndex].credentialsLen));
+        }
     }
 
     ReturnErrorOnFailure(ConnectivityMgrImpl().CommitConfig());
-    mSavedNetwork = mStagingNetwork;
+    memset(mSavedNetwork, 0, sizeof(mSavedNetwork));
+    memcpy(mSavedNetwork, mStagingNetwork, sizeof(mSavedNetwork));
     return CHIP_NO_ERROR;
 }
 
 CHIP_ERROR LinuxWiFiDriver::RevertConfiguration()
 {
-    mStagingNetwork = mSavedNetwork;
+    currentNetId = 0;
+    memset(mStagingNetwork, 0, sizeof(mStagingNetwork));
+    memcpy(mStagingNetwork, mSavedNetwork, sizeof(mStagingNetwork));
```

Comment:
Why do we need the memset, if we're going to memcpy anyway?

And presumably there should be a static_assert here that the two things have the same sizeof()?